### PR TITLE
SQL: Add RAND() function.

### LINKED ISSUE #8789: SQL RAND function
### Description

Add a `RAND()` function to Druid SQL.

### Motivation

It could be used for random sampling (e.g. `WHERE RAND() < 0.2`).

See also https://github.com/apache/incubator-druid/issues/8661.

### PR DESCRIPTION
**Pull request by gianm**
Fixes #8789, #8661.

### CODE REVIEW

**Comment by vogievetsky** [comment]
Does it fully fix #8661?

**Comment by gianm** [comment]
> Does it fully fix #8661?

I believe the other functions already exist (check the docs) so yes.

File: processing/src/main/java/org/apache/druid/query/expression/RandExprMacro.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    return "rand";
+  }
+
+  @Override
+  public Expr apply(List<Expr> args)
+  {
+    final DoubleSupplier randomGenerator;
+
+    if (args.isEmpty()) {
+      randomGenerator = () -> ThreadLocalRandom.current().nextDouble();
```
Comment (suneet-amp) [inline processing/src/main/java/org/apache/druid/query/expression/RandExprMacro.java]: If you wanted to make the test in ExpressionsTest.java easier, you could inject an annotated `Supplier<DoubleSupplier>` into this class something like

```
@Inject
RandExprMacro(@ThreadRandom Supplier<DoubleSupplier> localThreadRandomSupplier) {
    ...
}
```

As I'm typing this, it feels like overkill, so feel free to ignore

File: processing/src/main/java/org/apache/druid/query/expression/RandExprMacro.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+          throw new IAE("Function[%s] first argument must be a number literal");
+        }
+      } else {
+        throw new IAE("Function[%s] first argument must be a number literal");
+      }
+    } else {
+      throw new IAE("Function[%s] needs zero or one arguments", name());
+    }
+
+    class RandExpr extends ExprMacroTable.BaseScalarMacroFunctionExpr
```
[the remark this answers is not in the document]
Comment (gianm) [reply processing/src/main/java/org/apache/druid/query/expression/RandExprMacro.java]: It's sort of the same thing as an anonymous class, except it's not anonymous. I use them somewhat interchangeably. I'm not aware of any major advantage except that the class name is nicer in stack traces, debuggers, etc.

**Comment by gianm** [comment]
Actually, I just realized this is a bad idea, for two reasons.

1. The result of the random expression will be potentially cached at query time. Fixing this means adding an `isCacheable` method to all expressions.
2. At ingest time, if you use this in a `transformSpec`, it makes the results of ingestion nondeterministic. This will cause haywire with Kafka index task replicas. Fixing it probably requires creating a per-partition seed ahead of time on the supervisor and sharing it across tasks.

These issues can be overcome, but, it doesn't seem worth it at this time. So I will close this PR. Thanks @suneet-amp for taking a look anyway!

**Pull request closed without merging**